allocated-gpio/allocated-gpio.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * allocated-gpio.c
+ *
+ *  Created on: Apr 2, 2018
+ *      Author: Derrick Gibelyou
+ */
+#include "version.h"
+
+#include <linux/slab.h>
+#include <linux/gpio.h>
+#include <linux/of_gpio.h>
+#include <asm-generic/errno.h>
+#include <linux/sysfs.h>
+#include <linux/device.h>
+#include <linux/of_device.h>
+
+#define DEVICE_NAME "allocated-gpio"
+
+struct gpio_attribute
+{
+	struct device_attribute n;
+	u32 gpio;
+};
+
+struct gpio_driver_data
+{
+	struct gpio_attribute* attr_array;
+	struct attribute** attr_list;
+	struct attribute_group reg_attr_group;
+	int num_attrs;
+};
+
+
+ssize_t gpio_state_show(struct device *dev, struct device_attribute *attr, char *buf)
+{
+	struct gpio_driver_data *data = (struct gpio_driver_data*) dev->platform_data;
+	int ii = 0;
+	for (ii = 0; ii < data->num_attrs; ii++)
+	{
+		if (attr == &(data->attr_array[ii].n) )
+		{
+			int ret;
+			printk(KERN_DEBUG "Address match for %s\n", data->attr_array[ii].n.attr.name);
+			ret = gpio_get_value_cansleep(data->attr_array[ii].gpio);
+			return snprintf(buf, PAGE_SIZE, "%d\n", ret);
+		}
+	}
+	return snprintf(buf, PAGE_SIZE, "error\n");
+}
+
+ssize_t gpio_state_store(struct device *dev, struct device_attribute *attr, const char *buf, size_t size)
+{
+	struct gpio_driver_data *data = (struct gpio_driver_data*) dev->platform_data;
+	int ii = 0;
+	int value = buf[0]-'0';
+	if (value != 0 && value != 1)
+	{
+		printk(KERN_ERR "Invalid GPIO value: '%s'\n", buf);
+		return -EINVAL;
+	}
+
+	printk(KERN_DEBUG "Setting output %s to %c\n", attr->attr.name, buf[0]);
+	for (ii = 0; ii < data->num_attrs; ii++)
+	{
+		if (attr == &(data->attr_array[ii].n) )
+		{
+			printk(KERN_DEBUG "Address match for %s\n", data->attr_array[ii].n.attr.name);
+			gpio_direction_output(data->attr_array[ii].gpio, value);
+			return size;
+		}
+	}
+	return -EIO;
+}
+
+static struct of_device_id allocated_gpio_of_match[] = {
+		{.compatible = "allocated-gpio"},
+		{ /* end of table*/}
+};
+MODULE_DEVICE_TABLE(of, allocated_gpio_of_match);
+
+static void create_pin_attrs(struct platform_device *pdev)
+{
+	struct device_node *np = pdev->dev.of_node;
+	struct device_node *child;
+	struct gpio_driver_data *data;
+
+	int status;
+	size_t attr_size;
+	size_t attr_list_size;
+	u32 num_attrs = of_get_child_count(np);
+
+	data = (struct gpio_driver_data*)kzalloc(sizeof(struct gpio_driver_data), GFP_KERNEL);
+	if(PTR_ERR_OR_ZERO(data))
```

Comment:
It would be nice to start using vera++. I like that we're using tabs now. This line would break a rule for not having a space after a keyword.